.github/workflows/main.yml

```diff
@@ -1,6 +1,33 @@
 name: main
 
-on: [push, pull_request]
+on:
+  workflow_call:
+    inputs:
+      build-artifact-name:
+        required: true
+        type: string
+      build-artifact-path:
+        required: true
+        type: string
+      self-build:
+        required: false
+        type: number
+      pyodide-versions:
```

Comment:
I think we should not pass an array here and let people use a matrix strategy on the caller side. (same for browsers and runners)

```
jobs:
  job-test:
    strategy:
      matrix:
        pyodide-version: [0.21.0, 0.21.1]
    uses: pyodide/pytest-pyodide/....yml
    with:
      pyodide-version: ${{ matrix.pyodide-version }}
```

Reply:
The way it is written currently, you can pass a single value or an array for any of those inputs.

You can't just move the test-matrix generation outside the workflow entirely - internal generation and filtering of the test matrix needs to be in there really, because e.g. if you select a single pyodide version, you probably still want to test with all browsers and operating systems, and similarly if you select an OS, you probably still want to test different browsers, or if you select node, you're going to want to test with all node versions. If you're going to filter the matrices in main.yml anyway, it might as well have support for arrays in there, because that's a pretty trivial bit of code.

Reply:
I'm sorry, but I don't understand why filtering has to be done in main.yaml. Maybe it is because I am not very familiar with reusable action and I am not aware of its limitations. Could you check #48 and tell me if there is an issue that I wasn't aware of?

Reply:
I think it's because I wanted it to support simple use cases including:

1) Just call the pyodide-test main.yml and by default test on everything on latest version of pyodide. 

2) Call pyodide-test action and say I only want to run the macos tests, but test on everything that works there (so e.g. if chrome support for macos turned up, I would get it added). 

3) Or call pyodide-test action and select only chrome tests or only node tests, but all of them.

With your version, if you want to do a full test on all pyodide platforms, you need to copy the list from the build workflow into your project. By default it doesn't even test on any browsers. Then if pyodide releases support for a new platform, or some kind of new test configuration everyone needs to update their projects to support it.

Reply:
I understand your use cases, but I think we don't want to "test on everything" by default.
Maybe we can make two reusable workflows:

(1) A workflow that takes every parameter as inputs, so users can customize everything theirselves.
(2) A workflow that uses (1) and runs all tests by default (GHA recently [added support for nesting reusable workflows](https://github.blog/changelog/2022-08-22-github-actions-improvements-to-reusable-workflows-2/))

WDYT?

Reply:
Yep, I just woke up this morning thinking exactly this actually - split the python filter munging stuff into a separate workflow, make everything parameterised, but also include a testall.yaml which takes parameters like the current one.